### PR TITLE
Adds IC log file type, log viewer program, and door logging

### PR DESCRIPTION
**Pull request by nmajask**
# Document the changes in your pull request

Adds an IC log filetype to the game, which stores events and times and all that for IC events. Adds a program to view said file type, showing it in the same format as the robotact program. The point of these log files is to make machines have actual logs you can download, allowing for detectives to download and use them to figure out what happened. Currently the only machine with logging are airlocks, which log when someone opens/closes them and can be downloaded by hitting them with a mod pc when the panel is opened. Also includes some additional procs and stuff for files for use with other file types in the future.

image

# Wiki Documentation

At this point we should probably have a mod PC page. 

# Changelog
:cl:  
rscadd: Added a new log file type which is used to store events and a new program to view them
/:cl:

### CODE REVIEW

**Comment by ToasterBiome** [comment]
surveillance state

**Comment by maxion12345** [comment]
sleeper logs would be epic after this, maybe I'll take a crack at it somehow

**Comment by nmajask** [comment]
> sleeper logs would be epic after this, maybe I'll take a crack at it somehow

Logs arent too hard to add to machines, all you need to do is:

1. Set the has_logs var to true
2. Add a way to download logs, which is simple enough to do
3. Use the machinelog proc whenever you want something logged and using get_username to get the name of the user.

File: code/game/machinery/_machinery.dm [hunk 1]
```diff
@@ -633,3 +644,63 @@ Class Procs:
 			else
 				to_chat(user, "<span class='warning'>You fail to climb onto [src].</span>")
 	machineclimber = null
+
+// Adds an entry to the internal log
+
+/obj/machinery/proc/machinelog(var/string = "", var/time_override = null, var/overwright = TRUE)
```
Comment (adamsong) [inline code/game/machinery/_machinery.dm:650]: Overwrite is the variable name you are looking for

File: code/game/machinery/doors/airlock.dm [hunk 2]
```diff
@@ -1742,6 +1783,8 @@
 		to_chat(user, "The electrification wire has been cut")
 	else
 		set_electrified(MACHINE_DEFAULT_ELECTRIFY_TIME, user)
+		var/username = get_username(user)
+		machinelog(string = "Electrification enabled for [MACHINE_DEFAULT_ELECTRIFY_TIME][username ? " by [username]":""]")
```
Comment (adamsong) [inline code/game/machinery/doors/airlock.dm:1787]: Include units

File: code/game/machinery/doors/door.dm [hunk 3]
```diff
@@ -247,7 +257,7 @@
 	else if(I.tool_behaviour == TOOL_WELDER)
 		try_to_weld(I, user)
 		return 1
-	else if(!(I.item_flags & NOBLUDGEON) && user.a_intent != INTENT_HARM)
+	else if(!(I.item_flags & NOBLUDGEON) && user.a_intent != INTENT_HARM && I.GetID())
```
Comment (adamsong) [inline code/game/machinery/doors/door.dm:260]: Went this interfere with opening AA doors?

**Review by adamsong — CHANGES_REQUESTED** [review]

File: code/game/machinery/doors/airlock.dm [hunk 4]
```diff
@@ -1070,6 +1080,17 @@
 											span_notice("You cut through \the [src]'s outer grille."))
 						security_level = AIRLOCK_SECURITY_PLASTEEL_O
 					return
+		if(C.is_modular_computer())
+			if(isElectrified())
+				if(!justzap)
+					if(shock(user, 100))
+						justzap = TRUE
+						addtimer(VARSET_CALLBACK(src, justzap, FALSE) , 10)
```
Comment (Bibby0110) [inline code/game/machinery/doors/airlock.dm:1088]:
```suggestion
						addtimer(VARSET_CALLBACK(src, justzap, FALSE) , 1 SECONDS)
```

File: code/modules/modular_computers/file_system/logs.dm [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	size = entries.len * 0.1 + 0.5
+
+/datum/computer_file/log/proc/addentry(var/string = "", var/time_override = null, var/overwright = FALSE)
+	if(!string)
+		return
+	if(!entries)
+		entries = list()
+		return
+	var/isfull = entries.len >= max_entries
+	if(isfull)
```
Comment (Bibby0110) [inline code/modules/modular_computers/file_system/logs.dm:24]: just do the check in the if statement

File: code/modules/modular_computers/file_system/logs.dm [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		if(overwright)
```
Comment (Bibby0110) [inline code/modules/modular_computers/file_system/logs.dm:25]: Negate the if statement and do the return if !overwright,
then you drop the else and just put the entries.Cut(1,2) outside of the if

File: code/modules/modular_computers/file_system/logs.dm [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+			entries.Cut(1,2)
+		else
+			return
+	var/time
```
Comment (Bibby0110) [inline code/modules/modular_computers/file_system/logs.dm:29]:
```suggestion
	var/time station_time_stamp()
```

File: code/modules/modular_computers/file_system/logs.dm [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	if(time_override)
+		time = time_override
+	else
```
Comment (Bibby0110) [inline code/modules/modular_computers/file_system/logs.dm:32]: remove this and the below line

File: code/modules/modular_computers/file_system/logs.dm [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+		time = station_time_timestamp()
+	var/index = entries.len + 1
+	entries += list(list("time" = time, "content" = string, "index" = index))
+	calculate_size()
+
+/datum/computer_file/log/proc/clear(var/log = TRUE)
+	entries = list()
+	if(log)
```
Comment (Bibby0110) [inline code/modules/modular_computers/file_system/logs.dm:40]: wouldn't you want to calculate_size no matter if you logged the clearing or not

File: code/modules/modular_computers/file_system/logs.dm [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+/datum/computer_file/log/proc/findbytime(var/time, var/return_all = FALSE)
+	var/list/matches = list()
+	for(var/list/entry in entries)
+		if(entry["time"] == time)
+			if(return_all)
+				matches.Add(entry)
+			else
+				return entry
+	if(matches.len >= 0)
```
Comment (Bibby0110) [inline code/modules/modular_computers/file_system/logs.dm:53]: either you want just a > or you should just return it either way. can't get len < 0, and len = 0 is just empty

File: code/modules/modular_computers/file_system/logs.dm [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+	var/list/data = list()
+	var/string = ""
+	for(var/list/entry in entries)
+		if(return_string)
+			string += "[entry["time"]] - [entry["content"]]<br>"
+		else
+			data += "[entry["time"]] - [entry["content"]]"
+	if(return_string)
+		return string
+	else
```
Comment (Bibby0110) [inline code/modules/modular_computers/file_system/logs.dm:72]: no need for else

File: code/game/machinery/doors/airlock.dm [hunk 12]
```diff
@@ -544,10 +545,12 @@
 				updateDialog()
 			cont = TRUE
 	spawnPowerRestoreRunning = FALSE
+	machinelog(string = "Power has been restored")
```
Comment (Bibby0110) [inline code/game/machinery/doors/airlock.dm:548]: don't use named arguments for all of these.
(aka remove the string = )

File: code/game/machinery/_machinery.dm [hunk 13]
```diff
@@ -633,3 +644,63 @@ Class Procs:
...
+				var/mob/living/simple_animal/S = user
+				item = S.access_card
+				if(item && isidcard(item.GetID()))
+					id = item.GetID()
+	if(isitem(A))
+		var/obj/item/item = A
+		if(item.GetID())
+			id = item.GetID()
+	if(!username && id)
+		username = "[id.registered_name] - [id.assignment] [id.originalassignment == id.assignment?"":"([id.originalassignment])"]"
```
Comment (Bibby0110) [inline code/game/machinery/_machinery.dm:706]:
```suggestion
		username = "[id.registered_name] - [id.assignment] [id.originalassignment == id.assignment ? "" : "([id.originalassignment])"]" 
```

**Review by Bibby0110 — COMMENTED** [review]
what adam said, very cool idea

**Pull request closed without merging**